File: it.unibo.bls19Local/src/main/java/it/unibo/bls/listener/ButtonObserver.java
package it.unibo.bls.listener;

import it.unibo.bls.interfaces.IApplListener;
import it.unibo.bls.interfaces.IControlLed;
import java.util.Observable;

public class ButtonObserver  implements IApplListener {
private IControlLed controller;
	private int count   = 0;
//Factory method
	public static ButtonObserver createButtonListener(  ){
		return new ButtonObserver(   );
	}
	private ButtonObserver(  ){
		super();
 	}
	public void setControl( IControlLed ctrl ){
		this.controller = ctrl;
	}
	public int getNumOfClicks(){
		return count;
	}

	@Override  //from IApplListener IObserver -> Observer
	public void update( Observable source, Object state ){
		count++;
		//System.out.println("ButtonListener update | state=" + state + " controller=" + controller);
		controller.execute ( state.toString() );
	}
}

File: it.unibo.bls19Local/src/main/kotlin/it/unibo/bls/kotlin/listener/ButtonObserver.kt
package it.unibo.bls.kotlin.listener

import it.unibo.bls.kotlin.interfaces.IApplListener
import it.unibo.bls.kotlin.interfaces.IControl
import java.util.Observable

class ButtonObserver private constructor() : IApplListener {
    private var controller: IControl? = null
    private var count = 0
    override fun setControl(ctrl: IControl) {
        this.controller = ctrl
    }

    override fun getNumOfClicks(): Int {
        return count
    }

    override//from IObserver -> Observer
    fun update(source: Observable, state: Any) {
        count++
        //System.out.println("ButtonListener update | state=" + state + " controller=" + controller);
        controller!!.execute(state.toString())
    }

    companion object {
        //Factory method
        fun createButtonListener(): ButtonObserver {
            return ButtonObserver()
        }
    }
}
